### ISSUE TITLE
Rename `#[visited_node]` attr to `#[ast]`

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#4282]
At present we only have `#[visited_node]` attrs on types which are visited.

We will also need to add an attribute to all types which are part of AST but are *not* visited, for:

* #4284
* https://github.com/oxc-project/oxc/issues/4296
* AST transfer

Currently:

```rs
#[visited_node]
enum Expression<'a> { /* ... */ }
```

I propose:

```rs
#[ast(visit)]
enum Expression<'a> { /* ... */ }
```

or:

```rs
#[ast]
#[visit]
enum Expression<'a> { /* ... */ }
```

Types which are part of AST, but which are *not* visited would omit the `visit` part e.g.:

```rs
#[ast]
pub enum FunctionType { /* ... */ }
```

### Bikeshedding

We could call the attr something other than `#[ast]`. I considered `#[ast_node]` but:

1. That name will be familiar to some from SWC, and I don't think we should use it because these attrs do something very different - they're no-op markers for codegen, rather than macros generating a huge amount of trait impls and other code.
2. Some of these types (e.g. `FunctionType`) are not AST *nodes* per se.

The common factor between all these types is that they're part of the AST, so `#[ast]` seems descriptive.

But I'm not at all set on that name, if someone has a better suggestion.

cc @Boshen @rzvxa.

**Comment by rzvxa** [comment]
From day one I was a fan of this more than any other way so that's where I'm.

```rust
#[ast(visit)]
enum Expression<'a> { /* ... */ }
```

If you remember at first I was more in favor of naming it `ast_node` but your arguments against it(both here and https://github.com/oxc-project/oxc/pull/3815#issuecomment-2186739897) made me change my decision. `#[ast]` seems like a good choice👍

**Comment by rzvxa** [comment]
I had an idea for generated derives which I've never shared publicly. We might want to add an attribute such as `#[expand_derive(...)]` or `#[derived(...)]` to easily mark stuff like `Hash`, `PartialEq`, and now `CloneIn` derives.

It is a marker and wouldn't expand the derived macro base on the builtins(or proc_macro_derive), but it is a good way to telegraph that these are being derived for the said type.

**Comment by overlookmotel** [comment]
@Boshen and I discussed on video call just now. He's given his 👍 to this.

**Comment by rzvxa** [comment]
closed via #4289
